password/test/tasty/Validate.hs

```diff
@@ -0,0 +1,210 @@
+        PasswordTooLong _req actual -> genPassword actual (passPolicyCharSet policy)
+        NotEnoughReqChars category _req actual -> do
+          passwordLength <- genPasswordLength policy
+          let predicate = case category of
+                Uppercase -> isUpper
+                Lowercase -> isLower
+                Special   -> (\char -> char `elem` specialChars)
+                Digit     -> isDigit
+          let usableCharsets = T.filter (not . predicate) (passPolicyCharSet policy)
+          let requiredCharsets = T.filter predicate (passPolicyCharSet policy)
+          requiredChars <- replicateM actual (elements $ T.unpack $ requiredCharsets)
+          passwordText <- genPassword (passwordLength - actual) usableCharsets
+          shuffle (passwordText <> requiredChars)
+        InvalidChar chrs -> do
+          passwordLength <- genPasswordLength policy
+          passwordText <- genPassword (passwordLength - T.length chrs) (passPolicyCharSet policy)
+          -- Here, we make sure that the order of invalid characters are apporiate
+          -- or else the test will fail. For instance this will make the test fail
+          -- since the order of the characters are different
+          -- e.g. [InvalidChar "一二三"] /= [InvalidChar "三二一"]
+          shuffle (passwordText <> (T.unpack chrs)) `suchThat` (checkOrder policy (T.unpack chrs))
+        InvalidPasswordPolicy _invalidPolicy -> do
+          passwordLength <- genPasswordLength policy
+          genPassword passwordLength (passPolicyCharSet policy)
+      genPassword :: Int -> Text -> Gen String
+      genPassword num set = replicateM num (elements $ T.unpack set)
+      genPasswordLength :: PasswordPolicy -> Gen Int
+      genPasswordLength policy = do
+        let minLength = passPolicyMinimumLength policy
+        let maxLength = passPolicyMaximumLength policy
+        choose (minLength, maxLength)
+      checkOrder :: PasswordPolicy -> String -> String -> Bool
+      checkOrder policy invalidChars pass =
+        let filteredChars = filter (\c -> c `notElem` (T.unpack $ passPolicyCharSet policy)) pass
+         in invalidChars == filteredChars
+
+-- | Check if given 'InvalidReason' is valid
+isValidReason :: InvalidReason -> Bool
+isValidReason = \case
+  InvalidChar _ -> True
+  InvalidPasswordPolicy policy -> not $ isValidPasswordPolicy policy
+  PasswordTooLong required actual -> required < actual
+  PasswordTooShort required actual -> required > actual
+  NotEnoughReqChars _ required actual -> required > actual
+
+-- | 'PasswordPolicy' used for testing
+--
+-- Required characters are turned off so that it's much more easier to test.
+emptyPolicy :: PasswordPolicy
+emptyPolicy = PasswordPolicy 8 32 Nothing Nothing Nothing Nothing defaultCharSet
+
+upperCases :: String
+upperCases = ['A' .. 'Z']
+
+lowerCases :: String
+lowerCases = ['a' .. 'z']
+
+-- | Special characters
+specialChars :: String
+specialChars = " !\"#$%&'()*+,-./:;<=>?@[]^_`{|}~"
```

Comment:
is backslash (`\\`) not a valid special character?

Reply:
Yes, it is. I'll add them!